### ISSUE TITLE
Investigate data quality

### ISSUE THREAD

**Issue by dennybritz** [rpryzant/japanese_corpus#2]
We should figure out of what quality our aligned sentences are and how we can make them better. I think a simple way to do this is to just sample a couple (50 or so?) translation from each of the corpora and create a spreadsheet. I can go through the translations manually and mark the ones that are wrong, and why they are wrong (e.g. misaligned, garbage text, etc). That should give us an idea of how we can improve the data, and which corpora are the most noisy and need more work.

**Comment by rpryzant** [comment]
Section 4.4 of [this paper](http://www.lrec-conf.org/proceedings/lrec2016/pdf/621_Paper.pdf) outlines a formal method for alignment. It's called "JPO adequacy criterion" and was cooked up by the japanese patent office. The paper references a broken url ([HERE](http://www.jpo.go.jp/shiryou/toushin/chousa/tokkyohonyaku-hyouka.htm)) for official directions (in Japanese).

**Comment by rpryzant** [comment]
I trained my seq2seq with the same hyperparameters on similar-sized subsets of our data and a VI dataset I found. 


|      | 100k subset from subtitle corpus | 100k subset from IWSLT'15 English-Vietnamese |
|------|----------------------------------|----------------------------------------------|
| BLEU | 2.09                             | 18.44                                        |

This isn't super promising, but I changed some code between the time I started the EN-JA and EN-VI experiments, and something may have been broken at the onset. Plus, this isn't the best of comparisons. JA might just be intrinsically harder to learn. Now that we have ASPEC I'm going to redo this and compare against a 100k subset of ASPEC.

**Comment by rpryzant** [comment]
Same thing happened! Trained on a 100k subset of the data, and BLEUs in the 1-2 range. This isn't a language specific problem -- the same model gets > 10 BLEU on a 100k subset of ASPEC. 

For phrases like
```
ルール は 無 い ！
```
The model produces translations like,
```
<unk> が し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し し
```
Looks like it's converging nicely too:

![alt text](http://i.imgur.com/6zF8avF.png)

This is worrying!

**Comment by dennybritz** [comment]
Yeah, it could be that there is just too much noise in the data. I'll try though.

**Comment by rpryzant** [comment]
Yeah, I'm going to give one more shot and re-harvest a new cleaned corpus with a higher dictionary match threshold. 

If that doesn't work either, though, seems like the project is a wrap. Not everything works out...thats research. Thanks for tagging along :)

**Comment by dennybritz** [comment]
It's a bit strange, the model should learn something better. I don't think it's time to give up yet ;) I looked at the data manually and there definitely is signal in it. Maybe all that's needed is different hyperparameyers, more aggressive gradient clipping, etc. 

I'll run this data on the internal version and see if I get anything better.

**Comment by dennybritz** [comment]
Also, the data you posted doesn't have any dev/test datasets? I guess I'll just split it manually...

**Comment by rpryzant** [comment]
I hope you're right! 

FYI, you may want to impose a minimum word length on the corpus. I noticed that a lot of high-scoring phrases were one-word name utterances (because they consisted of one word which the dictionary could pick up on). Another reason why dictionary matching is suboptimal....

Nope I haven't made official dev/test splits yet. I've been randomly generating splits with [[THIS SCRIPT]](https://github.com/rpryzant/nmt/blob/master/phase_3_learning/data/generate_splits.py)

**Comment by rpryzant** [comment]
Smaller, (hopefully) higher-quality 500k subset of subtitle data. I sorted matches by score, threw out one-word phrases, then took the top ~500k.

Just started a training on on these data. 

* https://drive.google.com/file/d/0B6v0SLeq81aNSUttMWhNUndfM1k/view?usp=sharing

**Comment by dennybritz** [comment]
Ok, so I finished training on the ~1M corpus. Log perplexity only goes down to around 4, which means the data is quite noisy, but that was expected.

![image](https://cloud.githubusercontent.com/assets/403133/24116282/81b62350-0d63-11e7-99da-1b101814942c.png)

However, the model still seems to learn fine. I generated translation for the test set and here's what I got:

https://gist.github.com/dennybritz/be1b8a0199e29aa25159fb5b9e2d9023

Most of them are surprisingly good.

**Comment by rpryzant** [comment]
You're a wizard Denny. This is a big relief. I don't recognize a lot of those kanji but I agree, the translations look pretty good. 

I've noticed some of those things (duplicates, tags) myself. Another thing: there's visual queues like `(laughs)` that are in some of the subtitles. I'll work on fixing those problems in the complete 3.6M dataset, then regenerate all the subsets. I'll make some random splits too and update https://github.com/rpryzant/japanese_corpus/issues/5 with the results. We can start using those as our "official" datasets.

I suppose we're at another crossroads. Now that we know the dataset (sort of) works, let's decide on what to do with it. We've spoken about this a few times but we're better informed now. I think we could get a conference/workshop paper out of this without an unreasonable amount of extra work. I made https://github.com/rpryzant/japanese_corpus/issues/10 so that we can plan things out in one place

**Comment by rpryzant** [comment]
# dataset improvement steps

1. refactor my crawlers, improving reliability and readability (let no subfile go undownloaded!) 
2. refactor my parsers, improve tag cleaning, spellchecking, etc
3. think of better cleaners and execute on that

**Comment by dennybritz** [comment]
For some reason I'm having problems training on the corpus now, maybe it is because I regenerated the word pieces based on both ASPEC and (the noisy) subtitles. I'll try taining on the smaller not-so-noisy corpus.

**Comment by rpryzant** [comment]
Ok. I just got back and have scheduled all of the next two days on this. Hoping to get you some better data to work with.

**Comment by dennybritz** [comment]
Wow, that's great, thanks. I'm wondering why the first run on the subtitles was fine, but the subsequent ones seem to diverge. Maybe lucky initialization. I'm trying to train with SGD instead of Adam now, perhaps Adam results in gradients exploding when using noisy data, but that's just a wild guess.

**Comment by rpryzant** [comment]
Yeah best of luck. FYI, I trained on the 500k best-scoring phrases from the 1.6M dataset (data here: https://github.com/rpryzant/japanese_corpus/issues/4) and was able to get to a log perplexity of ~2 and word-level tokenization.

![data](http://i.imgur.com/3VwI6T3.png)

I'm going to start using your seq2seq and sentencepiece for any more experiments I run, though. For consistency's sake.

**Comment by dennybritz** [comment]
What parameters (optimizer, learning rate, gradient clipping, etc) did you use for that training run? I'm just trying it now, but SGD seems a lot more stable.

**Comment by rpryzant** [comment]
* batch size 128
* hidden lstm cells 512
* embedding size 256
* layers 2
* bilinear, global attention
* argmax decoding
* bidirectional encoder
* dropout 0.2
* adam 
* learning rate 0.0003 (no annealing)
* grad clipping at 5.0

**Comment by dennybritz** [comment]
These are pretty similar to mine. Given that I observe huge variations I'm starting to think more and more that this may be a bug in my code somewhere, perhaps due to some recent change. It could also be related to distributed training, so I'll try single-machine training just for the sake of comparison.

**Comment by dennybritz** [comment]
Confirmed that it's related to distributed training. If I train on a single machine it seems to work just fine using Adam. I'm not sure if this is a bug in my code related to distributed training, or something in Tensorflow.

**Comment by rpryzant** [comment]
Tensorflow hasn't quite ironed out distributed training yet...someone in my lab ran into a bunch of distributed training bugs last fall. 

Do you have access to a single machine with multiple GPUs? I'd hate for experiments to take more than a few days...

**Comment by dennybritz** [comment]
Hm, internally we're using distributed training for everything and it should be stable. I think it must be an issue with my code, maybe the data reading is off or something like that.

**Comment by dennybritz** [comment]
I do have access to machine with multiple GPUs, but I'm not sure how to best make use of that. I don't think the code efficiently works on multiple GPUs.

**Comment by dennybritz** [comment]
Surprisingly, this seems to be a somewhat common issue with Adam. There are two potential sources of the issue:

1. Worker replicas should be started with a delay as to not mess up the gradients in the beginning. I.e. start worker 2 after 1000 steps, worker 3 after 2000 steps, and so on.

2. Use Tensorflow's `SyncReplicasOptimizer` to synchronize gradients before applying them instead of applying them async. That makes it more stable.

**Comment by rpryzant** [comment]
Sounds good. Good luck.

On my end, I noticed two things:
1. Sub files have a ton of information in their filenames that we're ignoring
2. Sub files, even if they correspond to the same thing, may be shifted, so under the current alignment scheme they would result in completely invalid alignments that the system would think are great

So I'm trying out a new style of sub file alignment that I think will be much more powerful. I

1. do some fuzzy guessing/matching on the subfile name to align files
2. find the "first content caption" in both files, and shift all of their timestamps so that the entire file lines up. 
3. walk through both files, and as long as captions stay aligned, match them

Finding the first "content caption" is tough because you have to sort out what's a real caption vs what's msc stuff the subtitler threw in ("directed by...", "SUBS BY XXX", "yes/no", etc.). 

Right now I'm thinking of doing this by getting 
* lemmatized english definitions for all the content words of the first few japanese captions
* lemmatized content words for the first few english captions

Then I'll say that the "first matching caption" is the pair of sentences in the above lists that share the most lemmatized content words (nouns, verbs, adjectives).

**Comment by dennybritz** [comment]
That makes sense. For d-addicts we unfortunately don't have the original filenames (as far as I remember). We'd have to re-crawl it and save the filenames somewhere.

I think you should keep doing what you think is best, since you've spent much more time than me on this. I still feel like the vector-based matching is promising though. If done right, one should be able to make it invariant to constant shifts in time. It's basically a feature representation for two time series.

**Comment by dennybritz** [comment]
Basically, if you match the "shape" of the time series, you should be able to match any two subs pretty well, regardless of noise and constant shifts. E.g. something like wavelets or Gaussian kernels (I have no idea what I'm talking about).

**Comment by rpryzant** [comment]
Yeah I see what you mean...I agree with you, that's probably the right way to do things. Matching on filenames is pretty hacky. In any case, it seems like the harder problem is subtitle alignment, not file alignment, and right now I'm thinking it makes more sense to "read through" aligned files concurrently rather than blindly match on timestamps.

**Comment by dennybritz** [comment]
I fixed the issues with training. If you think it will be ready soon I will wait for the new data to run the experiments. Otherwise I can run them on the old one.

**Comment by rpryzant** [comment]
Great! That was quick. I've just about finished the aligner as well, but it will probably take a few days to run on the new crawl (it's huge!! By my calculations I'm expecting to get 6x the yield from kitsunikko) 

Why don't you use the old data to prove the idea works, then we can rerun it with the new data and (hopefully) improve our results?

**Comment by dennybritz** [comment]
Sounds good! If you have time it'd also be good to look at d-addicts again. I think we can get a lot more out of it (at least 2x) b/c my filtering was very simple

**Comment by rpryzant** [comment]
Out of curiosity what was the issue?

**Comment by rpryzant** [comment]
And yeah I'd like to take a look at d-addicts as well.

**Comment by dennybritz** [comment]
There are several issues:

1. I heard that Adam is generally somewhat unstable during distributed training because the asynchronous updates can mess up the gradient statistics it keeps track of.
2. Tensorflow recently made changes to their Adam optimizer implementation, adding momentum to sparse gradients (used for embeddings). As a result of that change my gradients seemed to explode with async updates. So I added gradient clipping of 0.1 for the embeddings.
3. It's generally a good idea to start workers with a delay when doing distributed training. E.g. start worker 1 after 1000 steps, worker 2 after 2000 steps, etc. That makes the async updates more stabl b/c otherwise you have 8 workers writing very different gradients in the beginning.

**Comment by rpryzant** [comment]
Cool. I think it makes intuitive sense that async updates would mess with adam. Adam uses a rmsprop-like leaky cache to store squared gradients. Stale/irrelevant gradients being added to that cache might throw the optimizer off. 

Also, FYI, the aligner is just about done, but it's looking like I'm going to have to parallelize across machines *and* threads if I want it to finish in less than a day. Working with Stanford IT to get my `apt-get` requirements on the right machines. Getting excited...almost ready for the big run!!

**Comment by dennybritz** [comment]
That sounds exciting, let's hope the run works!

I don't think it matters if the examples are sentences or not. But if the model never trains on sentences it will probably have a hard time generalizing from fragments to sentences. But as long as the test data is of that same distribution I think it doesn't really matter.

**Comment by rpryzant** [comment]
Ok. There's a mix of sentences and non-sentences so I'll forgo the segmenter.

**Comment by rpryzant** [comment]
Alright, everything's done and I've updated https://github.com/rpryzant/japanese_corpus/issues/4 with a new corpus from all this work.

I met my projections for kitsu (~6.1x), but interestingly, the number of subscene captions went **down** with the new aligner (from 1.3M to 1M). I don't think thats a big deal, though, because I think these matches are much higher quality.

On a whole, it isn't *that* much bigger than it was before (total size is ~2x), but I would expect it to be considerably higher quality. 

I'm going to focus on writing this stuff up in the paper for the next couple days.



Steps:

```

## recrawl subscene

python japanese_corpus/crawlers/subscene/v2/crawler.py urls/ja.txt > JA_LOG                                                                                                
python japanese_corpus/crawlers/subscene/v2/crawler.py urls/en.txt > EN_LOG



## recrawl kitsunikko

python japanese_corpus/crawlers/kitsunekko/v2/process_kitsu_dump.py



## reparse subscene and kitsunikko

japanese_corpus/parsers/subscene/final/parser_v4.py —num_threads 128 subscene_out out_en_subscene out_ja_subscene DONES_SUBSCENE
japanese_corpus/parsers/subscene/final/parser_v4.py —num_threads 128 kitsu_out out_en_kitsu out_ja_kitsu DONES_KITSU



## sort by cosine similarity, take top 800k from both, and rm consecutive dups

python subparse_sorter.py kitsu
tail -800000 out_ja_kitsu > OUT_JA_KITSU; tail -800000 out_en_kitsu > OUT_EN_KITSU; tail -800000 out_trans_kitsu > OUT_TRANS_KITSU
python rm_dups.py OUT_EN_KITSU OUT_JA_KITSU OUT_TRANS_KITSU

python subparse_sorter.py subscene
tail -800000 out_ja_subscene > OUT_JA_SUBSCENE; tail -800000 out_en_subscene > OUT_EN_SUBSCENE; tail -800000 out_trans_subscene > OUT_TRANS_SUBSCENE
python rm_dups.py OUT_EN_SUBSCENE OUT_JA_SUBSCENE OUT_TRANS_SUBSCENE



## combine and rm cross-source dups

cat daddicts/combined.en.cleaned kitsu/OUT_EN_KITSU_cleaned opensubtitles/combined.en subscene/OUT_EN_SUBSCENE_cleaned > total.en
cat daddicts/combined.ja.cleaned kitsu/OUT_JA_KITSU_cleaned opensubtitles/combined.ja subscene/OUT_JA_SUBSCENE_cleaned > total.ja

python rm_dups_unsorted.py total.en total.ja



## final round of cleaning + join ja/en lines

python final_clean.py total.en total.ja



## shuffle the combined lines, then split up again

cat combined.cleaned | shuf > combined.cleaned.shuffled
python split_combined.py combined.cleaned.shuffled
```

**Comment by dennybritz** [comment]
This is great, thanks! I think ~3M examples is more than enough. I still think we can get a lot more out of d-addicts but it's probably not high priority. This corpus should be fine for now! I'll start training on the new corpus too.
